packages/loader/container-loader/src/attachment.ts

```diff
@@ -0,0 +1,211 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { AttachState } from "@fluidframework/container-definitions";
+import { CombinedAppAndProtocolSummary } from "@fluidframework/driver-utils";
+import { ISnapshotTree } from "@fluidframework/protocol-definitions";
+import { assert } from "@fluidframework/core-utils";
+import { IDocumentStorageService } from "@fluidframework/driver-definitions";
+import { getSnapshotTreeAndBlobsFromSerializedContainer } from "./utils";
+import { ISerializableBlobContents } from "./containerStorageAdapter";
+import { IDetachedBlobStorage } from ".";
+
+/**
+ * The default state a newly created detached container will have.
+ * All but the state are optional and undefined, they just exist
+ * to make the union easy to deal with for both Detached types
+ */
+export interface DetachedDefaultData {
+	readonly state: AttachState.Detached;
+	readonly blobs?: undefined;
+	readonly summary?: undefined;
+	readonly redirectTable?: undefined;
+}
+
+/**
+ * This always follows DetachedDefaultData when there are
+ * outstanding blobs in the detached blob storage.
+ * The redirect table will get filled up to include data
+ * about the blobs as they are uploaded.
+ */
+export interface DetachedDataWithOutstandingBlobs {
+	readonly state: AttachState.Detached;
+	readonly blobs: "outstanding";
+	readonly summary?: undefined;
+	readonly redirectTable: Map<string, string>;
+}
+
+/**
+ * This state always follows DetachedDataWithOutstandingBlobs.
+ * It signals that all outstanding blobs are done being uploaded,
+ * so the container can move to the attaching state.
+ */
+export interface AttachingDataWithBlobs {
+	readonly state: AttachState.Attaching;
+	readonly blobs: "done";
+	readonly summary: CombinedAppAndProtocolSummary;
+}
+
+/**
+ * This always follows DefaultDetachedState when there are
+ * no blobs in the detached blob storage. Because there are
+ * no blobs we can immediately get the summary and transition
+ * to the attaching state.
+ */
+export interface AttachingDataWithoutBlobs {
+	readonly state: AttachState.Attaching;
+	readonly summary: CombinedAppAndProtocolSummary;
+	readonly blobs: "none";
+}
+
+/**
+ * The final attachment state which signals the container is fully attached.
+ * The baseSnapshotAndBlobs will only be enabled when offline load is enabled.
+ */
+export interface AttachedData {
+	readonly state: AttachState.Attached;
+	readonly baseSnapshotAndBlobs?: [ISnapshotTree, ISerializableBlobContents];
+}
+
+/**
+ * A union of all the attachment data types for
+ * tracking across all container attachment states
+ */
+export type AttachmentData =
```

Comment:
These are combo states, spanning both blob upload states and container attach states.  It would probably be conceptually simpler to split into two state machines that run sequentially - "upload detached blobs, if any" and "attach container using resultant redirect table, if any".  Esp. if the blob upload state can be encapsulated as suggested in the other comment.

Reply:
they are deeply coupled. the presence of blobs or not determines if we do zero byte upload, or summary upload, it also determines when we transition the runtime to attaching from detached. i think the code become much more complicated if they are decoupled.

Reply:
Even if you don't decouple/encapsulate, I still think it's worth considering 2 sequential machines here.

Reply:
i don't see a path here that doesn't add significant complexity to the code.